src/corels.cpp

```diff
@@ -175,79 +180,89 @@ int bbound(CacheTree* tree, size_t max_num_nodes, Queue* q, PermutationMap* p) {
     logger->incPrefixLen(0);
     // log record for empty rule list
     logger->dumpState();
-    while ((tree->num_nodes() < max_num_nodes) && !q->empty()) {
-        double t0 = timestamp();
-        std::pair<Node*, tracking_vector<unsigned short, DataStruct::Tree> > node_ordered = q->select(tree, captured);
-        logger->addToNodeSelectTime(time_diff(t0));
-        logger->incNodeSelectNum();
-        if (node_ordered.first) {
-            double t1 = timestamp();
-            // not_captured = default rule truthtable & ~ captured
-            rule_vandnot(not_captured,
-                         tree->rule(0).truthtable, captured,
-                         tree->nsamples(), &cnt);
-            evaluate_children(tree, node_ordered.first, node_ordered.second, not_captured, q, p);
-            logger->addToEvalChildrenTime(time_diff(t1));
-            logger->incEvalChildrenNum();
+}
 
-            if (tree->min_objective() < min_objective) {
-                min_objective = tree->min_objective();
-                if (verbosity >= 10)
-                    printf("before garbage_collect. num_nodes: %zu, log10(remaining): %zu\n", 
-                            tree->num_nodes(), logger->getLogRemainingSpaceSize());
-                logger->dumpState();
-                tree->garbage_collect();
-                logger->dumpState();
-                if (verbosity >= 10)
-                    printf("after garbage_collect. num_nodes: %zu, log10(remaining): %zu\n", tree->num_nodes(), logger->getLogRemainingSpaceSize());
-            }
-        }
-        logger->setQueueSize(q->size());
-        if (queue_min_length < logger->getQueueMinLen()) {
-            // garbage collect the permutation map: can be simplified for the case of BFS
-            queue_min_length = logger->getQueueMinLen();
-            //pmap_garbage_collect(p, queue_min_length);
-        }
-        ++num_iter;
-        if ((num_iter % 10000) == 0) {
+void bbound_loop(CacheTree* tree, Queue* q, PermutationMap* p) {
```

Comment:
I love that you separated this into a function -- the merging with our parallel code might be a bit tricky, but nice functional decomposition!

Reply:
Thank you :D